src/file_pipe_log/pipe.rs

```diff
@@ -32,6 +32,103 @@ struct ActiveFile<F: FileSystem> {
     writer: LogFileWriter<F>,
 }
 
+/// Collection of staled Files.
+struct RecycleFileCollection {
+    /// capacity of RecycleFiles
+    capacity: Option<usize>,
+    /// list of recycled files
+    fds: VecDeque<FileId>,
+}
+
+impl RecycleFileCollection {
+    pub fn new(capacity: Option<usize>) -> Self {
+        Self {
+            capacity,
+            fds: VecDeque::new(),
+        }
+    }
+
+    /// Check the RecycleFileCollection is open or not.
+    ///
+    /// Only when had the capacity been set with a valid value, the
+    /// `RecycleFileCollection` is valid.
+    pub fn valid(&self) -> bool {
+        self.capacity.is_some()
+    }
+
+    pub fn valid_to_push(&self, count: usize) -> bool {
+        match self.capacity {
+            Some(0) => true, // the capacity is infinite.
+            Some(p) => p >= count + self.fds.len(),
+            None => false,
+        }
+    }
+
+    /// Support update the capacity dynamically.
+    pub fn upd_capacity(&mut self, count: i64) {
```

Comment:
I prefer the related logic is put in-place, and keep recycle_files a simple collection. 

i.e. in `purge_to`:
```
let can_recycle = self.recycle_files.size() + self.files.size() < max_total_files;
```
In `rotate`:
```
if let Some(f) = self.recycle_files.first() {
  // open
} else {
  // create
}
```

Reply:
Ok, let me simplify the implementation here, as possible as i can.